libs/@blockprotocol/graph/src/codegen/preprocess/transform-type-names.ts

```diff
@@ -0,0 +1,154 @@
+} from "@blockprotocol/type-system/slim";
+import ts from "typescript";
+
+import { mustBeDefined } from "../../shared/util/must-be-defined.js";
+import {
+  typedEntries,
+  typedKeys,
+} from "../../shared/util/typed-object-iter.js";
+import { PreprocessContext } from "../context.js";
+import { generatedTypeSuffix } from "../shared.js";
+
+const typescriptKeywords = new Array(
+  ts.SyntaxKind.LastKeyword - ts.SyntaxKind.FirstKeyword,
+)
+  .fill(0)
+  .map((_, idx) => ts.tokenToString(ts.SyntaxKind.FirstKeyword + idx)!);
+
+const isTypescriptKeyword = (name: string) => {
+  return typescriptKeywords.includes(name);
+};
+
+/**
+ * Extracts the alphanumeric characters from the title and creates a Title Cased version that can be used as a
+ * TypeScript identifier
+ *
+ * @param title
+ */
+const generateValidTypeScriptIdentifierFromTitle = (title: string): string => {
+  /* @todo - Handle acronyms, we should do a non-case-sensitive match and then convert all the groups to lower-case */
+  // extract all letters and numbers from the title, and capitalise the start of each component
+  const titleCase = (title.match(/[a-zA-Z0-9]+/g) || [])
+    .map((word: string) => `${word.charAt(0).toUpperCase()}${word.slice(1)}`)
+    .join("");
+
+  if (!/[a-zA-Z]/.test(titleCase.charAt(0))) {
+    // if it starts with a number, append `T` to the start to make a valid type name
+    const typeName = `T${titleCase}`;
+    if (isTypescriptKeyword(typeName)) {
+      throw new Error(
+        `Generated type name "${typeName}" is a TypeScript keyword`,
+      );
+    }
+    return typeName;
+  } else {
+    if (isTypescriptKeyword(titleCase)) {
+      throw new Error(
+        `Generated type name "${titleCase}" is a TypeScript keyword`,
+      );
+    }
+    return titleCase;
+  }
+};
+
+/**
+ * **Rewrites** the `title`s of the types to be locally (per file) unique valid javascript identifiers.
+ *
+ * Clashes are dealt with per class of type, and then per Base Url, and then per revision.
+ */
+export const rewriteTypeTitles = (context: PreprocessContext) => {
+  context.logDebug("Rewriting type titles to be unique...");
+  /* eslint-disable no-param-reassign -- the point of this function is to mutate the types in place */
+  const typeNamesToTypes: {
+    dataType: Record<string, DataType[]>;
+    propertyType: Record<string, PropertyType[]>;
+    entityType: Record<string, EntityType[]>;
+  } = {
+    dataType: {},
+    propertyType: {},
+    entityType: {},
+  };
+
+  const typeNameOverrides = context.parameters.typeNameOverrides;
+
+  typedEntries(context.allTypes).forEach(([typeId, type]) => {
+    const override = typeNameOverrides[typeId];
+    const typeNameFromTitle = generateValidTypeScriptIdentifierFromTitle(
+      override ?? type.title,
+    );
+
+    if (override && typeNameFromTitle !== override) {
+      context.logWarn(
+        `Type name override of "${override}" for "${typeId}" isn't in Title Case, using "${typeNameFromTitle}" instead.`,
```

Comment:
Oof good catch. I think I had previously actually been converting it to Title Case and then allowing the JSON schema compiler to turn it into a type name as it wanted (which it did by turning it into PascalCase I think).

However I've decided to take greater control over this, as if we define the `title` as a valid typescript identifier the compiler doesn't seem to mutate it at all, which means we can use the rewritten `title` at other points in the pipeline _as_ the identifier, rather than needing to maintain some mapping.

I'll try and find a way to explain this slightly better in the code, and then update to replace Title Case with PascalCase in messages and variable names. Thanks!

Reply:
[Clean-up type title transformation](https://github.com/blockprotocol/blockprotocol/pull/1158/commits/9420d6ada452776cf2fc202e0a565cb1910af9cb)

I re-read the documentation of the top level function in this file and it does say that we rewrite the titles to valid JS identifiers so I think it's probably fine